src/server/page_server.cpp

```diff
@@ -244,6 +258,72 @@ page_server::connection_handler::prior_sender_sink_hook (std::string &&message)
   m_conn->push (page_to_tran_request::SEND_TO_PTS_LOG_PRIOR_LIST, std::move (message));
 }
 
+page_server::disconnect_handler::disconnect_handler ()
+  : m_terminate { false }
+{
+  m_thread = std::thread (&page_server::disconnect_handler::disconnect_loop, std::ref (*this));
+}
+
+page_server::disconnect_handler::~disconnect_handler ()
+{
+  assert (m_terminate);
+  if (m_terminate && m_thread.joinable ())
+    {
+      m_queue_cv.notify_one ();
+      m_thread.join ();
+    }
+
+  assert (m_disconnect_queue.empty ());
+}
+
+void
+page_server::disconnect_handler::disconnect (connection_handler_uptr_t &&handler)
+{
+  if (!m_terminate)
+    {
+      std::lock_guard<std::mutex> lockg { m_queue_mtx };
+      m_disconnect_queue.emplace (std::move (handler));
+      m_queue_cv.notify_one ();
+    }
+  else
+    {
+      // cannot ask for disconnect after termination
+      assert (false);
+    }
+}
+
+void
+page_server::disconnect_handler::terminate ()
+{
+  m_terminate = true;
+  m_queue_cv.notify_one ();
+}
+
+void
+page_server::disconnect_handler::disconnect_loop ()
+{
+  std::queue<connection_handler_uptr_t> disconnect_work_buffer;
+  while (!m_terminate)
+    {
+      {
+	std::unique_lock<std::mutex> ulock { m_queue_mtx };
+	m_queue_cv.wait (ulock, [this]
```

Comment:
How about waiting with timeout?

Reply:
What is the point of waiting with a timeout?
Whenever
1. work is pushed
2. terminate is set
the condition variable is notified.

Reply:
Is not about spinning. It is about calling wait_for instead wait. It is used for safety, to avoid infinite blocking

Reply:
Done